src/tests/end2end/IndexFormatTests.cpp

```diff
@@ -0,0 +1,268 @@
+                .SetIndexFormat(format)
+                .SetInputState(inputState)
+                .GetResult();
+        }
+};
+
+// Test that the Uint32 index format is correctly interpreted
+TEST_P(IndexFormatTest, Uint32) {
+    nxt::RenderPipeline pipeline = MakeTestPipeline(nxt::IndexFormat::Uint32);
+
+    nxt::Buffer vertexBuffer = utils::CreateFrozenBufferFromData<float>(device, nxt::BufferUsageBit::Vertex, {
+        -1.0f,  1.0f, 0.0f, 1.0f, // Note Vertices[0] = Vertices[1]
+        -1.0f,  1.0f, 0.0f, 1.0f,
+         1.0f,  1.0f, 0.0f, 1.0f,
+        -1.0f, -1.0f, 0.0f, 1.0f
+    });
+    // If this is interpreted as Uint16, then it would be 0, 1, 0, ... and would draw nothing.
+    nxt::Buffer indexBuffer = utils::CreateFrozenBufferFromData<uint32_t>(device, nxt::BufferUsageBit::Index, {
+        1, 2, 3
+    });
+
+    uint32_t zeroOffset = 0;
+    nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+        .BeginRenderPass(renderpass, framebuffer)
+        .BeginRenderSubpass()
+            .SetRenderPipeline(pipeline)
+            .SetVertexBuffers(0, 1, &vertexBuffer, &zeroOffset)
+            .SetIndexBuffer(indexBuffer, 0)
+            .DrawElements(3, 1, 0, 0)
+        .EndRenderSubpass()
+        .EndRenderPass()
+        .GetResult();
+
+    queue.Submit(1, &commands);
+
+    EXPECT_PIXEL_RGBA8_EQ(RGBA8(0, 255, 0, 255), renderTarget, 100, 100);
+}
+
+// Test that the Uint16 index format is correctly interpreted
+TEST_P(IndexFormatTest, Uint16) {
+    nxt::RenderPipeline pipeline = MakeTestPipeline(nxt::IndexFormat::Uint16);
+
+    nxt::Buffer vertexBuffer = utils::CreateFrozenBufferFromData<float>(device, nxt::BufferUsageBit::Vertex, {
+        -1.0f,  1.0f, 0.0f, 1.0f,
+         1.0f,  1.0f, 0.0f, 1.0f,
+        -1.0f, -1.0f, 0.0f, 1.0f
+    });
+    // If this is interpreted as uint32, it will have index 1 and 2 be both 0 and render nothing
+    nxt::Buffer indexBuffer = utils::CreateFrozenBufferFromData<uint16_t>(device, nxt::BufferUsageBit::Index, {
+        1, 2, 0, 0 ,0 ,0
+    });
+
+    uint32_t zeroOffset = 0;
+    nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+        .BeginRenderPass(renderpass, framebuffer)
+        .BeginRenderSubpass()
+            .SetRenderPipeline(pipeline)
+            .SetVertexBuffers(0, 1, &vertexBuffer, &zeroOffset)
+            .SetIndexBuffer(indexBuffer, 0)
+            .DrawElements(3, 1, 0, 0)
+        .EndRenderSubpass()
+        .EndRenderPass()
+        .GetResult();
+
+    queue.Submit(1, &commands);
+
+    EXPECT_PIXEL_RGBA8_EQ(RGBA8(0, 255, 0, 255), renderTarget, 100, 100);
+}
+
+// Test for primitive restart use vertices like in the drawing and draw the following
+// indices: 0 1 2 PRIM_RESTART 3 4 2. Then A and B should be written but not C.
+//             0
+//             |\
+//             |B \
+//             2---1
+//            /| C
+//          / A|
+//         4---3
```

Comment:
If, say, primitive restart were disabled, wouldn't this again go out of bounds?

If the restart index has to be 0xFFFFFFFFu, this isn't easily solvable for the Uint32 case (I think), but the Uint16 case could easily have a large enough vertex buffer that 0xFFFFu is a valid index.

Reply:
It would crash hopefully.